Allow clients to customize status bar styling

pg8wood opened the pull request.

mattczech commented on Duvet/SheetStatusBarStyleProvider.swift:
```diff
@@ -0,0 +1,10 @@
+
+/// Controls the source of truth for a `SheetViewController`'s status bar style.
+public enum SheetStatusBarStyleProvider {
+    /// Use the `SheetViewController`'s preferred status bar style.
+    case sheetViewController
```
In what case(s) would you pick this over `sheetItems`? Or is this mostly for backwards compatibility?

pg8wood replied:
Yup! The only purpose this serves is to use it as the default to preserve the existing behavior

mattczech replied:
This feels to me like a bug that we should fix across the board rather than a new feature that you have to opt into.

Unless you've customized the `preferredStatusBarStyle` of your view controller that you're showing in the sheet, it should return `.default` by default, right? So that should match the behavior of this `sheetViewController` option. So the only case we would break would be the case where the view controller had already overridden `preferredStatusBarStyle`, in which case I think we should honor that style automatically. What do you think about this?

pg8wood replied:
I'm cool with that. I was not sure if the existing functionality was intentional, so I didn't want to change the default behavior. In that case, I'll adapt this so that it's just a bug fix. 

I have one more bug fix and then we can cut a release bumping the patch version number

mattczech replied:
Sounds good, thanks!

pg8wood committed:
Respect sheet items' preferred status bar style

- Fix a bug where the SheetViewController would always show a dark status bar

pg8wood committed:
Add Status Bar example to sample app

pg8wood commented:
@mattczech Great call; just fixing it as a bug instead of introducing new public API for backwards compatibility has resulted in the SDK-side changes being pretty thin 😄

I've included a demo of the reference app changes as well: 
https://user-images.githubusercontent.com/16855664/181350718-199802a4-d654-40b5-9a11-fa0f525510c7.mov

mattczech commented on Duvet/SheetViewController.swift:
```diff
@@ -11,7 +11,15 @@ public class SheetViewController: UIViewController {
 
     /// An array of the sheet items being managed by the view controller.
-    public private(set) var sheetItems = [SheetItem]()
+    public private(set) var sheetItems = [SheetItem]() {
+        didSet {
+            setNeedsStatusBarAppearanceUpdate()
+        }
+    }
```
I'm seeing a slight flicker in the status bar when it's light and presenting a sheet that specifies a light status bar. It goes from white to black and back to white. I think `setNeedsStatusBarAppearanceUpdate()` is being called before the `currentSheetItem` is set.

I think we could drop the didSet observer here since we're calling  `setNeedsStatusBarAppearanceUpdate()` manually below as part of the transition. We could also maybe use `sheetItems.last` to get the `currentSheetItem` to avoid having to manually manage it. `sheetItems` represents the stack of sheets being displayed, with the last one currently being displayed.

pg8wood replied:
Nice catch!

mattczech commented on Examples/DuvetExample/DuvetExample/Application/ViewControllers/StatusBarViewController.swift:
```diff
@@ -0,0 +1,80 @@
+        supportedPositions: [.fittingSize]
+    )
+
+    private var statusBarStyle: UIStatusBarStyle = .lightContent {
+        didSet {
+            setNeedsStatusBarAppearanceUpdate()
```
Thanks for adding an example! Any way we could persist this during an app run? Then we'd also be able to see the transition when presenting a sheet too. Maybe a static var to maintain the value?

mattczech commented on Examples/DuvetExample/DuvetExample/Application/ViewControllers/StatusBarViewController.swift:
```diff
@@ -0,0 +1,80 @@
+import Duvet
+import UIKit
+
+enum StatusBarStyle: CaseIterable {
```
Can you add docs for this?

mattczech reviewed (commented):
Looking great! Just a few comments.

pg8wood committed:
Fix status bar flickering bug

Fix an issue where transitioning from a light status bar to light status bar could
briefly flash a dark status bar

pg8wood committed:
Persist modal preferred status bar style

Update docs

mattczech reviewed (approved):
Thanks @pg8wood, this looks great!
